### ISSUE TITLE
Deserialization of enums with name defined with different cases leads to `InvalidDefinitionException`: Multiple fields representing property

### ISSUE THREAD

**Issue by sbailliez** [FasterXML/jackson-databind#4409]
### Search before asking

- [X] I searched in the [issues](https://github.com/FasterXML/jackson-databind/issues) and found nothing similar.

### Describe the bug

I'm using a thirdparty client that is deserializing json where an enum is made of variations like RGBA and RGBa. When upgrading our dependencies from Jackson 2.14.1 to 2.16.1 this API started failing with

```
com.fasterxml.jackson.databind.exc.InvalidDefinitionException: Multiple fields representing property "rgba" ...
```

### Version Information

2.16.1 

Looks like the bug was introduced in 2.16.0 as it works with 2.15.4

### Reproduction

```java
    public static class Bug {
        public enum ColorMode {
            RGB, RGBa, RGBA
        }
        public ColorMode colorMode;
    }

    @Test
    public void enumFails() throws Exception {
        String json = "{ \"color_mode\": \"RGBa\"}";
        ObjectMapper mapper = new ObjectMapper();
        mapper.setPropertyNamingStrategy(PropertyNamingStrategies.SNAKE_CASE);
        mapper.disable(DeserializationFeature.FAIL_ON_UNKNOWN_PROPERTIES);
        Bug bug = mapper.readValue(json, Bug.class);
    }
``` 

### Expected behavior

This should succeed.

### Additional context

For this particular, I worked around this issue right now by patching the objectmapper created within the client and adding a custom deserializer for that enum:

```java
    public static class EnumDeserializer extends StdDeserializer<File.ColorMode> {
        public EnumDeserializer() {
            this(null);
        }

        public EnumDeserializer(Class<?> vc) {
            super(vc);
        }

        @Override
        public File.ColorMode deserialize(JsonParser p, DeserializationContext ctxt) throws IOException, JacksonException {
            JsonNode node = p.getCodec().readTree(p);
            String value = node.asText();
            return value != null ? File.ColorMode.valueOf(value) : null;
        }
    }
    
    // .... reflective access to find the objectmapper in the thirdparty client...
    mapper.registerModule(new SimpleModule("patch-enum")
         .addDeserializer(File.ColorMode.class, new EnumDeserializer()));

```

**Comment by cowtowncoder** [comment]
Thank you for reporting this, @sbailliez ! It definitely looks like bug.

**Comment by cowtowncoder** [comment]
Yes, looks like `PropertyNamingStrategy` is triggering the issue.

Related:

* #4302
* #4037 / #4040

**Comment by cowtowncoder** [comment]
Quick note: I haven't dig into why `"RGBA"` and '"RGBa"` are translated by convention, but you can work around the issue by adding `@JsonProperty` into one of conflicting values:

```
    enum ColorMode {
        RGB, RGBa,
        @JsonProperty("RGBA")
        RGBA
    }```

**Comment by cowtowncoder** [comment]
Ok the problem is that names do translate both to `rgba` as per `SNAKE_CASE_STRATEGY`:

https://github.com/FasterXML/jackson-databind/blob/2.17/src/main/java/com/fasterxml/jackson/databind/PropertyNamingStrategies.java#L171

So this is a problem that was not detected until 2.15 (I think) -- behavior in 2.16 is as expected going forward.

In case where naming strategy converts value into conflicts, annotation is needed, or possible `@EnumNaming` override: however, adding:

    @EnumNaming(EnumNamingStrategies.CamelCaseStrategy.class)

does not seem to help in this case (assuming it will not translate relevant values into separate ones)

**Comment by cowtowncoder** [comment]
Closing as wont-fix; solution is to apply annotation on one of conflicting values.

**Comment by sbailliez** [comment]
This is probably the first time that I see that the naming strategy was "supposed" to be applied to enums values. I thought there were expected to be serialized/deserialized as is and you would use valueOf or fromString to impact it. So I would vote for keeping backward compatibility. 

In my case this is a thirdparty library, so I cannot add annotations but will open a bug for trying to annotate it explicitly to make it compatible for all jackson versions.

**Comment by JooHyukKim** [comment]
Okay, so will modify PR #4414 to implement the new `EnumFeature` accordingly (probably later today).

But WRT version, shouldn't we have the feature provided in 2.16? So users upgrading from 2.15 can easily fix things.

**Comment by cowtowncoder** [comment]
No, as per SemVer no features may be added in patch releases: all patch releases should have same API -- additions may be done in minor releases.

But let's think this through first, before providing more solutions before understanding the problem.

What I'd want to know is the actual scope of problem; both wrt version in which behavior changes (wrt `PropertyNamingStrategy` applying to Fields of Enum types), and whether it actually changes serialization of regular Enums -- or is it just some special case. Given that we haven't had many reports, I wonder how commonly this actually changes behavior.

**Comment by cowtowncoder** [comment]
@sbailliez I may have spoken too soon on this change being intentional wrt `PropertyNamingStrategy`. It may well be unintentional change, due to somewhat aggressive rewriting of Enum handling in 2.16 (and some in 2.15 I think).

**Comment by JooHyukKim** [comment]
Did some digging and made tests against 2.15 and 2.16

- [2.15](https://github.com/JooHyukKim/jackson-databind/pull/5)🔗
    - PropertyNamingStrategy applied to both Serialization and Deserialization
- [2.16 ](https://github.com/JooHyukKim/jackson-databind/pull/6)🔗
    - PropertyNamingStrategy applied to ONLY deserialization

**Comment by sbailliez** [comment]
@JooHyukKim do you mean for 2.15 "NOT applied to both Serialization and Deserialization"?

**Comment by sbailliez** [comment]
@cowtowncoder I believe the reason there has not been any report related to this behavior in 2.16 is that it is pretty unconventional to have enums with values that have a case difference. In the API I refer to it is probably there for backward compatibility.

**Comment by cowtowncoder** [comment]
@sbailliez True, that makes it less common. Enums often have underscores and some naming conventions change those so... but either way, it is interesting how few problems have been reported

**Comment by cowtowncoder** [comment]
Ok so it's `EnumFormatShapeTest` that fails if skipping for all Enum types (wrt #2365 as per comment)

**Comment by cowtowncoder** [comment]
Ok. The trick is to check what `BasicBeanDescription.findExpectedFormat()` does, but from `POJOPropertiesCollector` -- and if we do that, I think ideally it'd be collected by `POJOPropertiesCollector` and then handed to `BasicBeanDescription` (to avoid multiple introspection calls).

But in short term (2.17) duplication would be fine.
This lookup is needed in 2.17 just for enum types, so we'd have something like:

```
    protected void _renameUsing(Map<String, POJOPropertyBuilder> propMap,
            PropertyNamingStrategy naming)
    {
        // _findFormatShape() needs to what `BasicBeanDescription.findExpectedFormat()`
        // (but need not take or use "defValue"; there's no defaulting)
        if (_type.isEnumType() && _findFormatShape() != Shape.OBJECT) {
            return;
        }
        ...
}
```

I might have time to look into this tomorrow, but if not I hope above helps.

**Comment by JooHyukKim** [comment]
Thanks @sbailliez  for reproduction and all!

**Comment by cowtowncoder** [comment]
Fixed or 2.16.2
